### ISSUE TITLE
ECDH using one-time ephemeral key

### ISSUE THREAD

**Issue by cfcs** [TLINDEN/pcp#14]
Paraphrasing from the README:
```
Please note the big difference to GPG though: both Alicia AND Bobby have to enter the passphrase for their secret key! That's the way CURVE25519 works: you encrypt a message using your secret key and the recipients public key and the recipient does the opposite, he uses his secret key and your public key to actually decrypt the message.
```
IMHO this is one of the few things PGP actually gets right: Unless you specifically CC yourself, you cannot read your own sent messages. This means an adversary who compromises your key at a later point can't decrypt all if your previously sent outgoing messages (only the incoming).

It can trivially be implemented for ECDH by generating a one-time keypair which is used for the key establishment between `(alice_onetime_x25519_sk ; bob_longterm_x25519_pk)` and sending `alice_onetime_x25519_pk` along with the message.
This is implemented in `libsodium` already, although it is confusingly named `crypto_box_seal`:
https://doc.libsodium.org/public-key_cryptography/sealed_boxes.html

To provide authenticity, you would include your `alice_longterm_ed25519_pk` in the message and sign the message using your `alice_longterm_ed25519_sk`.

Thoughts?

**Comment by TLINDEN** [comment]
I think I already implemented this: -A although not using crypto_box_seal() which didn't exist at the time.

**Comment by cfcs** [comment]
I'm having a hard time reading your code, but are you sure?
I think this is the code we're talking about:
https://github.com/TLINDEN/pcp/blob/bbdda67a6ec16fd46a2b465fff94164f7d339bac/src/encryption.c

It seems to me this only happens if `anon` is set to `1`, in which case the call to `pcp_rec_new()` in https://github.com/TLINDEN/pcp/blob/4e5c27c58cb867be2c915cee99683ad9eed089b3/libpcp/crypto.c#L307 would use the new anonymous keypair to do the signing. Or? Maybe I overlooked something?

**Comment by TLINDEN** [comment]
Well, I read the documentation to crypto_box_seal(). It doesn't attach or include a longtime pk to the message, also it doesn't sign it with the longterm pk. Instead it explicitly states:
```
Without knowing the secret key used for a given message, the sender cannot decrypt its own message later. And without additional data, a message cannot be correlated with the identity of its sender.
```
So, my implementation might be totally different, but it follows the same rules. Yes, it is possible to also sign the message in anonymous mode using the ephemeral pk for this, which is nonsense. However, using the longterm pk to sign the message would undermine anonymity, the message could be identifiable.

- Tom

**Comment by cfcs** [comment]
Okay, so here are two solutions that solve these problems. I'm going to call the first ANONYMOUS, and the second PRIVATE (because privacy and anonymity are not exactly the same, and it's fair to assume that some individuals would like their identities to remain private and known only by the author and the recipients).

- ANONYMOUS: Encrypted message from an anonymous author. ECDH between ephemeral key and recipients. Signing is pointless since the recipients can't verify the author; the MAC tag is only useful for verifying that the message wasn't accidentally altered or cut off, and doesn't protect against attacker that seek to replace the message. List of recipients pks is included to prevent attackers that cannot read the message from cutting off recipients (not perfect since other recipients can reencrypt and alter this).
```
context_info = list of recipient pks
payload = context_info || plaintext_message

Sent:
    ephemeral_pk || nonce || encrypt_and_mac(key = symmetric_msg_key, msg = payload)
    For each recipient:
        encrypt(msg = symmetric_msg_key, key = ecdh(sk = ephemeral_sk, pk = recipient))
```

- PRIVATE: Encrypted message from an anonymous author whom the recipient may or may not known, but who seeks to reveal their identity only to the recipient. ECDH between ephemeral key and recipients. Signature is sent along with the message (but encrypted). The list of recipients is included (like email CC) to prevent (all) attackers from cutting off recipients, provided the author pk is known to the recipients.
```
context_info = author_pk || nonce || symmetric_msg_key || list of recipient public keys
payload = sign(secret_key = author_sk, msg = (context_info || plaintext_message))

Sent:
    ephemeral_pk || nonce || encrypt_and_mac(key = symmetric_msg_key, msg = payload)
    For each recipient:
        encrypt(msg = symmetric_msg_key, key = ecdh(sk = ephemeral_sk, pk = recipient))
```

To decrypt this, you would compute the shared key for each recipient slot in turn and attempt to validate the mac.

I hope this highlights what I meant, of course there are nicer constructions, especially for larger messages where you'll probably want to be able to see if you can read a message without having to compute a mac on the whole thing for each recipient slot (can be solved by having a header that's mac'ed separately), etc.

**Comment by TLINDEN** [comment]
Well, that's going into the wrong direction. I'm not encrypting "messages" or "files", but streams. Since to the program it is unknown when a stream ends, it cannot produce separate outputs for each recipient (well it could in theory but it would be a mess).

From the docs:

    Formal format description, asymetric encrypted files:
    
     +-----------------------------------------------------------+
     | Field         Size      Description                       |
     +-------------+--------+------------------------------------+
     | Type        |      1 | Filetype, 5=ASYM, 23=SYM, 6=ANON   |
     +-------------|--------|------------------------------------+
     | Anon PUB *  |     32 | anon pubkey, only used with type 6 |
     +-------------|--------|------------------------------------+
     | Len R *     |      4 | Number of recipients         (*)   |
     +-------------|--------|------------------------------------+
     | Recipients *|   R*72 | C(recipient)|C(recipient)... (*)   |
     +-------------|--------|------------------------------------+
     | Encrypted   |      ~ | The actual encrypted data          |
     +-------------|--------|------------------------------------+
    
    *) not included when doing symetric encryption.
    
    Recipient field format:
    
     +---------------------------------------------------------+
     | Field         Size      Description                     |
     +-------------+--------+----------------------------------+
     | Nonce       |     24 | Random Nonce, one per R          |
     +-------------|--------|----------------------------------+
     | Cipher      |     48 | S encrypted with PK or R         |
     +-------------|--------|----------------------------------+
    
    R is generated using B<crypto_box()> with the senders
    secret key, the recipients public key and a random nonce.
    
    Pseudocode:
    
     R = foreach P: N | crypto_box(S, N, P, SK)
     L = len(R)
     T = 5
     write (T | L | R)
     foreach I: write (N | crypto_secret_box(I, N, S))
    
    where P is the public key of a recipient, SK is the senders
    secret key, R is the recipient list, L is the number of recipients,
    T is the filetype header, I is a block of input with a size
    of 32k, N is a nonce (new per block) and S the symmetric key.
    
    If using anonymous encryption, the sender generates a ephemeral
    key pair, uses the secret part of it to generate R. The public
    part will be included with the output (right after the file type.
    In this mode a recipient is not required to have the public key
    of the sender.
    
    The encrypted output maybe Z85 encoded. In this case the Z85
    encoding will be done blockwise with blocks of 16k bytes. The
    decoded content inside will be as described above.

So, in short, in order to see if a recipient is able to decrypt, she iterates over the recipient list and tries to decrypt each item using her own sk and the anonymous pk (or, if not anonymous, one of the pk's she has installed). If one item decrypts she uses it as symmetric key to decrypt the stream which follows. Also note, that I do not include recipient pk's in any case. The only key material sent, if any, is the pk in anonymous mode, which is ephemeral anyway.

So, to support what you need (as I understand it) would be to sign such a message using the long term pk and keep everything else as is.


- Tom

**Comment by cfcs** [comment]
Yep, pretty much.
